src/features/ExtensionCommands.ts

```diff
@@ -382,23 +387,54 @@ export class ExtensionCommandsFeature implements IFeature {
+    private saveFile(saveFileDetails: ISaveFileDetails): Thenable<EditorOperationResponse> {
 
-        let promise: Thenable<EditorOperationResponse>;
-        if (this.findTextDocument(this.normalizeFilePath(filePath))) {
-            promise =
-                vscode.workspace.openTextDocument(filePath)
-                    .then((doc) => {
-                        if (doc.isDirty) {
-                            doc.save();
-                        }
-                    })
-                    .then((_) => EditorOperationResponse.Completed);
-        } else {
-            promise = Promise.resolve(EditorOperationResponse.Completed);
+        // If the file to save can't be found, just complete the request
+        if (!this.findTextDocument(this.normalizeFilePath(saveFileDetails.filePath))) {
+            return Promise.resolve(EditorOperationResponse.Completed);
         }
 
-        return promise;
+        // If no newFile is given, just save the current file
+        if (!saveFileDetails.newPath) {
+            return vscode.workspace.openTextDocument(saveFileDetails.filePath)
+                .then((doc) => {
+                    if (doc.isDirty) {
+                        doc.save();
+                    }
+                })
+                .then((_) => EditorOperationResponse.Completed);
+        }
+
+        // Otherwise we want to save as a new file
+
+        // First turn the path we were given into an absolute path
+        // Relative paths are interpreted as relative to the original file
+        const newFileAbsolutePath = path.isAbsolute(saveFileDetails.newPath) ?
+            saveFileDetails.newPath :
+            path.resolve(path.dirname(saveFileDetails.filePath), saveFileDetails.newPath);
+
+        // Create an "untitled-scheme" path so that VSCode will let us create a new file with a given path
+        const newFileUri = vscode.Uri.parse("untitled:" + newFileAbsolutePath);
+
+        // Now we need to copy the content of the current file,
+        // then create a new file at the given path, insert the content,
+        // save it and open the document
+        return vscode.workspace.openTextDocument(saveFileDetails.filePath)
+            .then((oldDoc) => {
+                return vscode.workspace.openTextDocument(newFileUri)
+                    .then((newDoc) => {
+                        return vscode.window.showTextDocument(newDoc, 1, false)
+                            .then((editor) => {
+                                return editor.edit((editBuilder) => {
+                                    editBuilder.insert(new vscode.Position(0, 0), oldDoc.getText());
+                                })
+                                .then(() => {
+                                    return newDoc.save()
+                                        .then(() => EditorOperationResponse.Completed);
+                                });
+                            });
+                    });
+            });
```

Comment:
I think instead of this scary nested code, we could use something like `Promise.all` and fire all the ones we can in that and then use the data.

Promise.all notes:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all